File: code/567.go
/*
 * @Author: lmio
 * @Date: 2023-02-05 23:56:17
 * @LastEditTime: 2023-02-19 23:29:10
 * @FilePath: /leetcode/code/567.go
 * @Description:567.字符串的排列
 */
package code

func CheckInclusion(s1, s2 string) bool {
    l1, l2 := len(s1), len(s2)
    if l1 > l2 {
        return false
    }
    cnt := [26]int{}
    for _, ch := range s1 {
        cnt[ch - 'a']--
    }

    left := 0
    for right, ch := range s2 {
        index := ch - 'a'
        cnt[index]++
        for cnt[index] > 0 {
            cnt[s2[left] - 'a']--
            left++
        }
        if right - left + 1 == l1 {
            return true
        }
    }
    return false
}

File: code/567.py
'''
Author: lmio 2091319361@qq.com
Date: 2023-08-25 22:15:50
LastEditors: lmio 2091319361@qq.com
Description: 567. 字符串的排列
'''

from collections import Counter, defaultdict


class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        m = set(s1)
        k = len(s1)
        if len(s2) < k:
            return False
        cnts1 = Counter(s1)
        cnts = defaultdict(int)
        def contain() -> bool:
            for ch in m:
                if cnts[ch] != cnts1[ch]:
                    return False
            return True
        left = 0
        for i in range(k):
            cnts[s2[i]] += 1
        if contain():
            return True
        for right in range(k, len(s2)):
            cnts[s2[left]] -= 1
            cnts[s2[right]] += 1
            left += 1
            if s2[right] in m and contain():
                return True
        return False

s = Solution()
print(s.checkInclusion("ab", "a"))